keps/sig-node/4381-dra-structured-parameters/kep.yaml

```diff
@@ -40,8 +41,21 @@ feature-gates:
       - kube-controller-manager
       - kube-scheduler
       - kubelet
+  - name: DRAAdminAccess
+    components:
+      - kube-apiserver
+    # This feature gate will be moved into a separate KEP later
+    # because it remains in alpha. It controls whether the "AdminAccess"
+    # field can be set. If set, other components support it
+    # because the field might have been set when enabled.
+    # Therefore other components don't need to and shouldn't
```

Comment:
So, this field should be omitted from the beta1 API, unless the feature gate is set, shouldn't it?

Also, clients should also have the feature gate (kube-scheduler in particular), and only take action if the feature gate is set. It doesn't matter if the value was set while the gate was enabled; the feature should not have an effect when the gate is disabled.

Reply:
A while back @thockin started a discussion about how we handle feature gates. I can't find it again now, but if I remember correctly, it boiled down to the observation that we handle feature gates inconsistently. We either:
- Put all code related to it behind an `if` check.
- Check at the boundary, typically the apiserver.

I am using the second approach here because I think the first one leads to additional error scenarios. For example:
- A user was able to request admin access (either by using 1.31, or 1.32 with feature gate enabled) when deploying a monitoring solution.
- Feature gate gets disabled before pod scheduling.
- Scheduler treats those claims like normal claims.
- They now prevent using hardware that was meant for production workloads. To the user it is not obvious that this happened because they see the monitoring solution as deployed with admin access in the request.

Reply:
Found it: https://docs.google.com/document/d/1roVAHyF7eWZAccmCKw7MXYUNgx4BCDSXF2IMS8h10oY/edit?resourcekey=0-x6Tw2qz1SpCIPhbec6Qa2A

> It would be very bad if their stored object started failing validation or worse, their cluster started acting badly.

It's this "acting badly" that happens when ignoring the admin access field because of a feature gate check in code which is supposed to honor that field.

> We do not have good guidelines about whether implementation logic should itself be gate-controlled or not.

In particular, "only take action if the feature gate is set" is not a general guideline.

As explained in the doc, not putting code behind an if check comes with risk that the code is faulty and then gets invoked despite the feature gate being off. This is a real problem which has no good answer in the general case.

Coming back to admin access: the code which would have to be put behind an if check is pretty short. I think the risk that it's faulty is lower than the problems that arise when not executing it, so I stand by my proposal to always honor the field if it was set.

Reply:
So, if the code in kube-scheduler that handles admin access is causing it to crash loop, then doing it "only at the boundary" means that turning off the feature gate won't fix the crashing. You have to turn off the feature gate, and then delete or update the claim with admin access. I don't like that.

Reply:
Having said that, I don't care enough to fight over this. If you want the if checks, then I can add them.

Reply:
The alternative to "ignore the field" is to explicitly cause allocation to fail when it is set and the feature gate is off.

Reply:
@pohly let's do that - your last suggestion. Refuse to allocate if the field is set. In particular, this is the lowest risk option from a security standpoint, while still avoiding an "incorrect" (allocated NOT in admin mode) outcome.

For posterity: https://kubernetes.slack.com/archives/CPNHUMN74/p1728317533313819

Reply:
Done.